packages/ui/src/main.ts

```diff
@@ -0,0 +1,110 @@
+import show, { type WalletProviderWithStoreVersion } from "./modal"
+import Bowser from "bowser"
+import sn, { type StarknetWindowObject } from "get-starknet-core"
+
+type StoreVersion = "chrome" | "firefox" | "edge"
+
+function getStoreVersionFromBrowser(): StoreVersion {
+  const browserName = Bowser.getParser(window.navigator.userAgent)
+    .getBrowser()
+    .name?.toLowerCase()
+  switch (browserName) {
+    case "firefox":
+      return "firefox"
+    case "microsoft edge":
+      return "edge"
+    default:
+      return "chrome"
+  }
+}
+
+export interface ConnectOptions {
+  modalMode?: "alwaysAsk" | "canAsk" | "neverAsk"
+  modalTheme?: "light" | "dark" | "system"
+  storeVersion?: StoreVersion
+}
+
+export interface DisconnectOptions {
+  clearDefaultWallet?: boolean
+}
+
+export const connect = async ({
+  modalMode = "canAsk",
+  storeVersion = getStoreVersionFromBrowser(),
+  modalTheme,
+}: ConnectOptions = {}): Promise<StarknetWindowObject | null> => {
+  const enableWithVersion = async (wallet: StarknetWindowObject | null) => {
+    if (!wallet) {
+      return null
+    }
+    return sn.enable(wallet, { starknetVersion: "v4" }).catch(() => null)
+  }
+
```

Comment:
the `"neverAsk"` (aka silent connect flow) attempt should be executed first, overriding all other states, 
i.e. `lastWallet` could be set but not preauthorized so the user will be shown with a popup

Reply:
e.g. lastWallet & preAuthorized could be mutually exclusive (there is a chosen wallet, but the user decided to reject its connection request, so now we have a chosen wallet instance which isn't preAuthorized)

Reply:
this doesn't handle neverAsk + multiple preAuthorized wallets

Reply:
I think it will be much simpler to handle it case by case.
```
if (neverAsk) {
    // at this point preAuthorized is sorted/included/excluded/shuffled properly 
    const wallet = preAuthorized[0];
    return enableWithVersion(wallet)
}
...

const forcePopup = isCurrentlyConnected || "alwaysAsk" || !lastWallet;
if (!forcePopup) {
    // handle the complex logic
}

return show({...})
```

Reply:
I refactored the part to follow the pattern suggested.

It now returns each case early, but still shares as much code as possible

Reply:
tnx, pls see my suggestion -

```
Index: packages/ui/src/main.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/ui/src/main.ts b/packages/ui/src/main.ts
--- a/packages/ui/src/main.ts	(revision c2469723a8c56ee357fcca7282e11d5a8bbc1131)
+++ b/packages/ui/src/main.ts	(date 1663058117523)
@@ -55,35 +55,35 @@
     ...restOptions,
   })
 
-  const lastWallet = await sn.getLastConnectedWallet()
-  if (modalMode !== "alwaysAsk") {
-    if (
-      lastWallet &&
-      preAuthorizedWallets.some((w) => w.id === lastWallet.id)
-    ) {
-      return enableWithVersion(lastWallet)
-    }
-  }
-
   const defaultWallet = await sn.getDefaultWallet()
+  const lastWallet = await sn.getLastConnectedWallet()
+
   if (modalMode === "neverAsk") {
-    if (
-      defaultWallet &&
-      preAuthorizedWallets.some((w) => w.id === defaultWallet.id)
-    ) {
-      return enableWithVersion(defaultWallet)
-    }
+    const wallet =
+      preAuthorizedWallets.find((w) => w.id === defaultWallet?.id) ??
+      preAuthorizedWallets.find((w) => w.id === lastWallet?.id) ??
+      preAuthorizedWallets[0] // at this point pre-authorized is already sorted
 
-    if (preAuthorizedWallets.length === 1) {
-      return enableWithVersion(preAuthorizedWallets[0])
-    }
-
-    return null
+    // return `wallet` even if it's null/undefined since we aren't allowed
+    // to show any "connect" related UI
+    return enableWithVersion(wallet)
   }
 
   const installedWallets = await sn.getAvailableWallets(restOptions)
-  if (modalMode === "canAsk" && installedWallets.length === 1) {
-    return enableWithVersion(installedWallets[0])
+  if (
+    modalMode === "canAsk" &&
+    // we return/display wallet options once per first-dapp (ever) connect
+    (defaultWallet || lastWallet)
+  ) {
+    const wallet =
+      preAuthorizedWallets.find((w) => w.id === defaultWallet?.id) ??
+      preAuthorizedWallets.find((w) => w.id === lastWallet?.id) ??
+      installedWallets.length === 1
+        ? installedWallets[0]
+        : undefined
+    if (wallet) {
+      return enableWithVersion(wallet)
+    } // otherwise fallback to modal
   }
 
   const discoveryWallets = await sn.getDiscoveryWallets(restOptions)
```

I find it more readable, plmk what u think.

Reply:
@janek26 do you find this code problematic?

Reply:
no :) merged